projects/plugins/jetpack/functions.opengraph.php

```diff
@@ -352,81 +352,287 @@ function jetpack_og_get_image( $width = 200, $height = 200, $deprecated = null )
-		 *
-		 * @since 10.4
-		 *
-		 * @param string $str Default Open Graph image alt text.
-		 */
-		$image['alt_text'] = apply_filters( 'jetpack_open_graph_image_default_alt_text', '' );
+	return \Automattic\Jetpack\Publicize\Social_Image_Generator\Templates::TEMPLATES;
+}
+
+/**
+ * Get a social image token from Social Image Generator.
+ *
+ * @since $$next-version$$
+ *
+ * @param string $site_title The site title.
+ * @param string $image_url The image URL.
+ * @param string $template The template to use.
+ *
+ * @return string|WP_Error The social image token, or a WP_Error if the token could not be generated.
+ */
+function jetpack_og_get_social_image_token( $site_title, $image_url, $template ) {
+	// Let's check if we have a cached token.
+	$cache_key      = wp_hash( $site_title . $image_url . $template );
+	$transient_name = 'jetpack_og_social_image_token_' . $cache_key;
+	$cached_token   = get_transient( $transient_name );
+
+	if ( ! empty( $cached_token ) ) {
+		return $cached_token;
 	}
 
-	return $image;
+	/**
+	 * Filter the social image token for testing purposes.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param string|WP_Error|null $token The token to return, or null to use default behavior.
+	 */
+	$token = apply_filters( 'jetpack_og_get_social_image_token', null );
+	if ( null !== $token ) {
+		return $token;
+	}
+
+	if ( ! function_exists( '\Automattic\Jetpack\Publicize\Social_Image_Generator\fetch_token' ) ) {
+		return new WP_Error( 'jetpack_og_get_social_image_token_error', __( 'Social Image Generator is not available.', 'jetpack' ) );
+	}
+
+	$token = \Automattic\Jetpack\Publicize\Social_Image_Generator\fetch_token( $site_title, $image_url, $template );
+
+	// If we have a token, cache it for a day.
+	if ( ! is_wp_error( $token ) ) {
+		set_transient( $transient_name, $token, DAY_IN_SECONDS );
+	}
+
+	return $token;
+}
+
+/**
+ * Generate and create a fallback social image.
+ *
+ * @since $$next-version$$
+ *
+ * @param array  $representative_image The representative image of the site.
+ * @param string $template The template to use.
+ *
+ * @return array The source ('src'), 'width', and 'height' of the image.
+ */
+function jetpack_og_generate_fallback_social_image( $representative_image, $template ) {
+	$site_title     = get_bloginfo( 'name' );
+	$fallback_image = array(
+		'src'    => $representative_image['src'],
+		'width'  => $representative_image['width'],
+		'height' => $representative_image['height'],
+	);
+
+	// Ensure that we use a valid template.
+	if (
+		! in_array(
+			$template,
+			jetpack_og_get_available_templates(),
+			true
+		)
+	) {
+		$template = 'edge';
+	}
+
+	// Let's generate the token matching the image..
+	$token = jetpack_og_get_social_image_token( $site_title, $representative_image['src'], $template );
+
+	if ( is_wp_error( $token ) ) {
+		return $fallback_image;
+	}
+
+	// Build the image URL and return it.
+	return array(
+		'src'    => sprintf(
+			'https://s0.wp.com/_si/?t=%s',
```

Comment:
@gmjuhasz I see that we use Jetpack Redirects to generate the URL for SIG, since the beginning:
https://github.com/Automattic/jetpack/pull/29118/commits/17507a2568ce0212d55884e579bd5f0a3890b860

I searched on p2 to try to understand the reason behind it, but couldn't find any conversation about this. Do you happen to have some background on this? Do you think I should do the same here?

Reply:
To be honest I'm not sure. In the UI panel we use the base URL, because of CORS issues:

https://github.com/Automattic/jetpack/blob/2736a60c996b57be346ec3812e1e47795d24eb48/projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js#L40-L46

I think it could have been made like that for tracking with Jetpack redirects? 🤔